src/util/finalizable_processor.rs

```diff
@@ -0,0 +1,140 @@
+use crate::scheduler::gc_work::ProcessEdgesWork;
+use crate::scheduler::{GCWork, GCWorker};
+use crate::util::ObjectReference;
+use crate::MMTK;
+use std::marker::PhantomData;
+
+/// A special processor for Finalizable objects.
+// TODO: we should consider if we want to merge FinalizableProcessor with ReferenceProcessor,
+// and treat final reference as a special reference type in ReferenceProcessor.
+#[derive(Default)]
+pub struct FinalizableProcessor {
+    /// Candidate objects that has finalizers with them
+    candidates: Vec<ObjectReference>,
+    /// Index into candidates to record where we are up to in the last scan of the candidates.
+    /// Index after nursery_index are new objects inserted after the last GC.
+    nursery_index: usize,
+    /// Objects that can be finalized. They are actually dead, but we keep them alive
+    /// until the binding pops them from the queue.
+    ready_for_finalize: Vec<ObjectReference>,
+}
+
+impl FinalizableProcessor {
+    pub fn new() -> Self {
+        Self {
+            candidates: vec![],
+            nursery_index: 0,
+            ready_for_finalize: vec![],
+        }
+    }
+
+    pub fn add(&mut self, object: ObjectReference) {
+        self.candidates.push(object);
+    }
+
+    fn get_forwarded_finalizable<E: ProcessEdgesWork>(
+        e: &mut E,
+        object: ObjectReference,
+    ) -> ObjectReference {
+        e.trace_object(object)
+    }
+
+    fn return_for_finalize<E: ProcessEdgesWork>(
+        e: &mut E,
+        object: ObjectReference,
+    ) -> ObjectReference {
+        e.trace_object(object)
+    }
+
+    pub fn scan<E: ProcessEdgesWork>(&mut self, e: &mut E, nursery: bool) {
+        let start = if nursery { self.nursery_index } else { 0 };
+
+        // We should go through ready_for_finalize objects and keep them alive.
+        // Unlike candidates, those objects are known to be alive. This means
+        // theoratically we could do the following loop at any time in a GC (not necessarily after closure phase).
+        // But we have to iterate through candidates after closure.
+        self.candidates.append(&mut self.ready_for_finalize);
+
+        for reff in self
+            .candidates
+            .drain(start..)
+            .collect::<Vec<ObjectReference>>()
```

Comment:
This code copy `candidates` to another vector, clear the old storage, iterate the new copy to construct a new `candidates` vector. I think it is possible to rewrite the loop using [`Vec::drain_filter`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter) here to reduce the copy of this vector.

Reply:
Thanks. I will fix this.

Reply:
`drain_filter` is not stable yet, and also I'm not sure how `drain_filter` can be efficiently used to get the a range of elements because the filter is applied on individual elements.

Reply:
I quickly tested those two. It seems `drain_filter()` is better. I will do this refactoring. `drain_filter()` is not stable, but it doesn't seem to be an issue for us.

```rust
#![feature(test)]
#![feature(drain_filter)]

extern crate test;
fn main() {
    println!("Hello, world!");
}

pub fn drain(x: &mut Vec<usize>) -> Vec<usize> {
    let mut ret = vec![];
    for i in x.drain(0..).collect::<Vec<usize>>() {
        if i % 2 == 0 {
            x.push(i);
        } else {
            ret.push(i);
        }
    }

    ret
}

pub fn drain_filter(x: &mut Vec<usize>) -> Vec<usize> {
    let mut ret = vec![];
    for i in x.drain_filter(|x| *x % 2 == 0) {
        ret.push(i);
    }
    ret
}

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    fn data() -> Vec<usize> {
        (0..100000).collect()
    }

    #[bench]
    fn bench_drain(b: &mut Bencher) {
        b.iter(|| { let mut x = data(); drain(&mut x) });
    }

    #[bench]
    fn bench_drain_filter(b: &mut Bencher) {
        b.iter(|| { let mut x = data(); drain_filter(&mut x) });
    }
}
```

Result
```
yilin@rat:~/tests/drain-filter/drain-bm/src$ cargo +nightly bench
    Finished bench [optimized] target(s) in 0.00s
     Running /home/yilin/tests/drain-filter/drain-bm/target/release/deps/drain_bm-bbd2ae3ecd458477

running 2 tests
test tests::bench_drain        ... bench:     419,649 ns/iter (+/- 5,855)
test tests::bench_drain_filter ... bench:     207,699 ns/iter (+/- 2,138)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out; finished in 0.67s
```

Reply:
Right. But `drain(start..).collect::<Vec<ObjectReference>>()` gives you a vector containing the slice `[start:]`. However, `drain_filter(predicate)` gives you a vector containing all elements satisfy the predicate. I'm not sure how to translate this.

Reply:
Umm, that's right. It would depend on what function you apply for each item. In our case, we trace objects for each item. Probably for now, I will just leave it as it is.